### PR TITLE
[fga] allow toggling centralizedPermissions on/off

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description

Allow to toggle the `centralizedPermissions` featureFlag off for users, so they will get migrated/updated again when the featureFlag is tunred on later.

## Related Issue(s)

Fixes EXP-376

#### Preview status

<ul>
        <li><b>🏷️ Name</b> - se-toggle-fga</li>
        <li><b>🔗 URL</b> - <a href="https://se-toggle-fga.preview.gitpod-dev.com/workspaces" target="_blank">se-toggle-fga.preview.gitpod-dev.com/workspaces</a>.</li>
        <li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
        <li><b>📦 Version</b> - </li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -151,11 +151,16 @@ jobs:
           MYSQL_TCP_PORT: 23306
         ports:
           - 23306:23306
+      redis:
+        image: redis
+        ports:
+          - 6379:6379
```
Comment (svenefftinge) [inline .github/workflows/build.yml:157]: We now run a redis service similar to how we run the test db in GHA.

File: components/dashboard/src/data/featureflag-query.ts [hunk 2]
```diff
@@ -12,7 +12,6 @@ import { useCurrentUser } from "../user-context";
 import { useCurrentOrg } from "./organizations/orgs-query";
 
 const featureFlags = {
-    start_with_options: false,
```
Comment (svenefftinge) [inline components/dashboard/src/data/featureflag-query.ts:15]: A zombie feature flag that has long time not been used

File: components/server/src/authorization/relationship-updater.ts [hunk 3]
```diff
@@ -53,41 +61,40 @@ export class RelationshipUpdater {
                 fromVersion: user?.additionalData?.fgaRelationshipsVersion,
                 toVersion: this.version,
             });
-            // TODO run in distributed lock
-            // return this.mutex.using([`fga-migration-${user.id}`], 2000, async () => {
-            const before = new Date().getTime();
+            return this.mutex.using([`fga-migration-${user.id}`], 2000, async () => {
```
Comment (svenefftinge) [inline components/server/src/authorization/relationship-updater.ts:64]: now that redis work in GHA build, we can again run this code in a distributed lock

File: components/server/package.json [hunk 4]
```diff
@@ -17,15 +17,15 @@
...
-    "test:db": ". $(leeway run components/gitpod-db:db-test-env) && yarn start-services && mocha --opts mocha.opts './**/*.spec.db.js'  --exclude './node_modules/**'",
+    "test:db": "cleanup() { echo 'Cleanup started'; yarn stop-spicedb; }; trap cleanup EXIT; . $(leeway run components/gitpod-db:db-test-env) && yarn start-spicedb && mocha --opts mocha.opts './**/*.spec.db.js'  --exclude './node_modules/**'",
     "start-services": "yarn start-testdb && yarn start-redis && yarn start-spicedb",
     "stop-services": "yarn stop-redis && yarn stop-spicedb",
     "start-testdb": "if netstat -tuln | grep ':23306 '; then echo 'Mysql is already running.'; else leeway run components/gitpod-db:init-testdb; fi",
     "start-spicedb": "leeway run components/spicedb:start-spicedb",
     "stop-spicedb": "leeway run components/spicedb:stop-spicedb",
-    "start-redis": "if netstat -tuln | grep ':6379 '; then echo 'Redis is already running.'; else docker run --rm --name test-redis -p 6379:6379 -d redis; fi",
+    "start-redis": "if lsof -i:6379 -n | grep LISTEN; then echo 'Redis is already running.'; else docker run --rm --name test-redis -p 6379:6379 -d redis; fi",
```
Comment (akosyakov) [inline components/server/package.json:28]: clean up to use netstat

File: components/server/src/authorization/relationship-updater.ts [hunk 5]
```diff
@@ -42,6 +44,12 @@ export class RelationshipUpdater {
             },
         });
         if (!fgaEnabled) {
+            if (user.additionalData?.fgaRelationshipsVersion === undefined) {
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply components/server/src/authorization/relationship-updater.ts]: Good catch! I should have added a test in the first place. Added one now.

**Review by akosyakov — APPROVED** [review]
code changes make sense, not sure how to verify (beside reviewing unit tests)

File: components/server/src/authorization/relationship-updater.ts [hunk 6]
```diff
@@ -42,6 +44,12 @@ export class RelationshipUpdater {
             },
         });
         if (!fgaEnabled) {
+            if (user.additionalData?.fgaRelationshipsVersion !== undefined) {
+                log.info({ userId: user.id }, `User has been removed from FGA.`);
+                // reset the fgaRelationshipsVersion to undefined, so the migration is triggered again when the feature is enabled
+                AdditionalUserData.set(user, { fgaRelationshipsVersion: undefined });
+                return await this.userDB.storeUser(user);
```
Comment (geropl) [inline components/server/src/authorization/relationship-updater.ts:51]: I wonder if it make sense to remove the old relations in SpiceDB at this point...
I bet you thought about it, and would love to hear your line of thought. :pray:

Comment (svenefftinge) [reply components/server/src/authorization/relationship-updater.ts:51]: The relationship-updater takes care of that once the user is enabled again. In the meantime the relationships have no relevance.

**Pull request merged**